### PR TITLE
Improve error handling for badging API

### PR DESCRIPTION
**Pull request by marcoscaceres**
#### 978ccadf6e0baf1d12fb5d43a1b515ffb45b20e6
<pre>
Improve error handling for badging API
<a href="https://bugs.webkit.org/show_bug.cgi?id=254193">https://bugs.webkit.org/show_bug.cgi?id=254193</a>
rdar://106969959

Reviewed by Brady Eidson.

Adds InvalidStateErrors for promise rejections.
Now checks if the document is fully active before setting the badge, and rejects if it&apos;s not.

* LayoutTests/imported/w3c/web-platform-tests/badging/META.yml: Added.
* LayoutTests/imported/w3c/web-platform-tests/badging/badge-error.https-expected.txt: Added.
* LayoutTests/imported/w3c/web-platform-tests/badging/badge-error.https.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/badging/badge-success.https-expected.txt: Added.
* LayoutTests/imported/w3c/web-platform-tests/badging/badge-success.https.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/badging/idlharness.https.any-expected.txt: Added.
* LayoutTests/imported/w3c/web-platform-tests/badging/idlharness.https.any.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/badging/idlharness.https.any.js: Added.
(idl_array.self.GLOBAL.isWorker):
* LayoutTests/imported/w3c/web-platform-tests/badging/idlharness.https.any.worker-expected.txt: Added.
* LayoutTests/imported/w3c/web-platform-tests/badging/idlharness.https.any.worker.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/badging/non-fully-active.https-expected.txt: Added.
* LayoutTests/imported/w3c/web-platform-tests/badging/non-fully-active.https.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/badging/w3c-import.log: Added.
* LayoutTests/imported/w3c/web-platform-tests/workers/interfaces/WorkerUtils/navigator/007.js:
* LayoutTests/platform/gtk/fast/dom/navigator-detached-no-crash-expected.txt:
* LayoutTests/platform/mac-wk1/TestExpectations:
* LayoutTests/platform/mac-wk2/fast/dom/navigator-detached-no-crash-expected.txt:
* LayoutTests/platform/wpe/fast/dom/navigator-detached-no-crash-expected.txt:
* Source/WebCore/page/Navigator.cpp:
(WebCore::Navigator::setAppBadge):
* Source/WebCore/page/WorkerNavigator.cpp:
(WebCore::WorkerNavigator::setAppBadge):
* Tools/WebKitTestRunner/TestOptions.cpp:
(WTR::TestOptions::defaults):

Canonical link: <a href="https://commits.webkit.org/262237@main">https://commits.webkit.org/262237@main</a>
</pre>

https://github.com/WebKit/WebKit/commit/c8209b971aca3babec1f04c77394ad2cff3a9f0b

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/892 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/26/builds/916 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/951 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/1266 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/788 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/880 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/23/builds/966 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/1001 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/996 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/900 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/15/builds/837 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/831 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/1201 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/892 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/826 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/832 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/7/builds/807 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/852 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/1857 "Passed tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/30/builds/848 "Built successfully") | [❌ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/17/builds/801 "3 failures") | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/4/builds/795 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/837 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/240 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/31/builds/862 "Built successfully") | | | 
| | | | |

### CODE REVIEW

**Comment by marcoscaceres** [comment]
I could use a bit of guidance with getting the tests to run... I'm a bit unsure what I should set UnifiedWebPreferences.yaml to restrict testing to EWS and locally?

This API is only exposed in certain situations (e.g., only for Home Screen Web Apps) so it can't be enabled by default for WebKit. 

Any suggestions?

**Comment by marcoscaceres** [comment]
I ended up using Tools/WebKitTestRunner/TestOptions.cpp to enable the tests. It seems to work as expected. 

Hope that's correct!

**Comment by marcoscaceres** [comment]
Failing test on mac-wk1 seems unrelated.

File: Source/WebCore/page/Navigator.cpp [hunk 1]
```diff
@@ -388,16 +388,24 @@ void Navigator::setAppBadge(std::optional<unsigned long long> badge, Ref<Deferre
...
         return;
     }
 
+    auto* document = frame->document();
+    if (document) {
+        if (!document->isFullyActive()) {
+            promise->reject(InvalidStateError);
+            return;
+        }
+    }
```
Comment (beidson) [inline Source/WebCore/page/Navigator.cpp]: if (document && !document->isFullyActive()) { ... }

**Review by beidson — APPROVED** [review]

**Pull request merged**